### PR TITLE
This is the parity generation/rebuild using AVX-512 for x86-64.

### PR DESCRIPTION
**Pull request by rdolbeau**
avx512f should work on all AVX512 hardware, since it only uses Foundation instructions.

avx512bw should be faster on hardware supporting the AVW512BW extension.

### CODE REVIEW

**Comment by rdolbeau** [comment]
This can be tested in userland with the Intel SDE <https://software.intel.com/en-us/articles/intel-software-development-emulator>, e.g.

sde-external-7.49.0-2016-07-07-lin/sde64 -skx -- cmd/raidz_test/raidz_test -S -v

This will test both avx512f & avx512bw, since Skylake server ("skx") supports both.

raidz_test -S seems OK on Knights Landing hardware as well (avx512f only).

File: module/zfs/vdev_raidz_math_avx512f.c [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+ * This uses a pair of AVX2 pshufb to emulate the missing AVX512 pshufb.
+ * AVX512BW has the full pshufb
+ * To get VEX pshufb (AVX2, supported in KNL) instead of EVEX pshufb
+ * (AVX512BW, not supported on KNL, probably also requiring AVX51VL
+ * since we use a 256 bits version), all registers in parameters to
+ * pshufb must be among ymm0-ymm15, since only EVEX can encore
+ * ymm16-ymm31
+ * This is a bit hackish, but short of encoding the instruction in
+ * binary, how do we force theuse of AVX2 pshufb ?
+ * Not that the other way round (forcing AVX512) is easy, just encode
```
Comment (rlaager) [inline module/zfs/vdev_raidz_math_avx512f.c]: Should this be "Note" instead of "Not"?

File: module/zfs/vdev_raidz_math_avx512f.c [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+/*
+ * This uses a pair of AVX2 pshufb to emulate the missing AVX512 pshufb.
+ * AVX512BW has the full pshufb
+ * To get VEX pshufb (AVX2, supported in KNL) instead of EVEX pshufb
+ * (AVX512BW, not supported on KNL, probably also requiring AVX51VL
+ * since we use a 256 bits version), all registers in parameters to
+ * pshufb must be among ymm0-ymm15, since only EVEX can encore
+ * ymm16-ymm31
+ * This is a bit hackish, but short of encoding the instruction in
+ * binary, how do we force theuse of AVX2 pshufb ?
```
Comment (rlaager) [inline module/zfs/vdev_raidz_math_avx512f.c]: This needs a space in "theuse".

File: include/sys/vdev_raidz_impl.h [hunk 3]
```diff
@@ -141,6 +141,12 @@ extern const raidz_impl_ops_t vdev_raidz_ssse3_impl;
 #if defined(__x86_64) && defined(HAVE_AVX2)	/* only x86_64 for now */
 extern const raidz_impl_ops_t vdev_raidz_avx2_impl;
 #endif
+#if defined(__x86_64) && defined(HAVE_AVX512F)	/* only x86_64 for now */
```
Comment (rlaager) [inline include/sys/vdev_raidz_impl.h]: I realize this just copying the existing comment, but why "for now"? Isn't this inherently x86_64 only?

File: module/zfs/vdev_raidz_math_avx512f.c (hunk 2 above)
Comment (rdolbeau) [inline module/zfs/vdev_raidz_math_avx512f.c]: fixed

File: module/zfs/vdev_raidz_math_avx512f.c (hunk 1 above)
Comment (rdolbeau) [inline module/zfs/vdev_raidz_math_avx512f.c]: fixed

File: include/sys/vdev_raidz_impl.h (hunk 3 above)
Comment (rlaager) [inline include/sys/vdev_raidz_impl.h]: Are these extensions available when the processor is running in 32-bit mode? If not, I'd rather see the comment dropped.

**Comment by rlaager** [comment]
Thanks for this work, by the way! I'm glad to see you taking on this sort of performance enhancement. I'm not an assembly programmer, so I'm sorry I can't be of more help in the review process.

File: include/sys/vdev_raidz_impl.h (hunk 3 above)
Comment (rdolbeau) [inline include/sys/vdev_raidz_impl.h]: Actually, they are - to my surprise :-)

From the "Intel® 64 and IA-32 Architectures Software Developer’s Manual" chapter 15 "Programming with Intel® AVX-512" they are. Although (15.1.2), "The number of available vector registers in 32-bit mode is still 8."

Comment (rlaager) [inline include/sys/vdev_raidz_impl.h]: Hmm, that is surprising. Well, it sounds like you've addressed all my concerns. Thanks!

Comment (ironMann) [inline include/sys/vdev_raidz_impl.h]: Yes, 8 available regs makes it hard to implement reconstruction (hard but not impossible). However, with ABD change reg pressure is somewhat lower, so if there's demand, 32bit can be revisited.

**Comment by behlendorf** [comment]
@rdolbeau can you address the `ccpcheck` warning.

http://build.zfsonlinux.org/builders/Amazon%202015.09%20x86_64%20%28STYLE%29/builds/164/steps/shell_3/logs/stdio

**Comment by rdolbeau** [comment]
> can you address the ccpcheck warning.
@behlendorf : Don't think so; this is a side-effect of the original SSE/AVX2 implementations. The number of arguments to the macros is used to determine the codepath, but the arguments themselves are passed to both variants - so sometimes there's too many or too few registers in the dead code. ... I think.

**Comment by behlendorf** [comment]
We're going to need to do something to silence the warning.

**Comment by rdolbeau** [comment]
@behlendorf Actually, it seems there is a difference with AVX2 - bad copy/paste, somehow. I'll try a fix (even if it changes nothing, it makes the diff between avx2 and avx512bw even smaller so more readable).

**Comment by ironMann** [comment]
@rdolbeau I've fixed something similar before, also flagged by `cppcheck`. I think [this](https://github.com/zfsonlinux/zfs/commit/c9187d867fee3972de48b71762407ae7dabb2563#diff-e17f41e70ef8f94916a9ebc22c70c609R50) does it.

**Comment by rdolbeau** [comment]
@ironMann Yep, that's what I've changed as well.
Unrolling more (gen_pq only in 512f) is a bit better:
```
cat /proc/spl/kstat/zfs/vdev_raidz_bench
14 0 0x01 -1 0 2690442302864469 2690453468983938
implementation   gen_p           gen_pq          gen_pqr         rec_p           rec_q           rec_r           rec_pq          rec_pr          rec_qr          rec_pqr         
original         277003462       31240228        18945130        277795538       23577988        4257231         12107637        2256264         2270898         1554264         
scalar           163513131       101422407       45466079        151400471       46720036        35430586        31103188        26346315        18844013        12770143        
sse2             788361648       311128866       170627031       774064341       147113774       115213303       90084548        81485197        47513866        20413132        
ssse3            787946117       377631746       170683213       774363968       158942113       125889240       87363735        78389721        46327029        33918767        
avx2             1393670156      790035858       381938724       1362427763      329758942       277027183       182058269       167448571       99221631        72745484        
avx512f          2240697713      1146759619      528378352       2164437372      360094996       287970665       202761667       182349557       104817759       75214419        
fastest          avx512f         avx512f         avx512f         avx512f         avx512f         avx512f         avx512f         avx512f         avx512f         avx512f  
```

**Comment by rdolbeau** [comment]
@behlendorf Seems fixed now (at least , the Style checks is OK).

**Comment by ironMann** [comment]
@rdolbeau Currently available avx512f hardware doesn't seem to be geared towards running zfs, or anything like that (albeit it's possible with KNL chips). I don't know what exactly plan is, but server-like CPUs will have 512BW.

When developing fletcher4 for avx512f, I found that the cpu (Xeon Phi 7210) handles vector 8bit shuffles very poorly. By switching to implementation that uses 3x more shift instruction, I was able to achieve ~4x throughput. I'm wondering if similar can be done here as well, for multiplication macro. I implemented this for SSE2 already, where shuffle is not available.

**Comment by rdolbeau** [comment]
@ironMann Isn't the shuffle in fletcher only used for byte reversal? Then replacing by shifts is surely a win on KNL. But in this, it's a data-dependent shuffle, more difficult to replace.

And yes, KNl is a HPCCPU not really a GPCPU; I doubt anyone will make a Lustre OSS with it. But maybe someone might want to run a local ZFS some day, who knows :-)

**Comment by ironMann** [comment]
> Isn't the shuffle in fletcher only used for byte reversal? Then replacing by shifts is surely a win on KNL. But in this, it's a data-dependent shuffle, more difficult to replace.

@rdolbeau That's right. But, for vector multiplication in SSE2 I created separate fixed method (using mul2 and xors) for each multiplicand, thus removing data dependency. There's overhead in fn call/ret but it's still a big performance win over scalar. Also, only few of these multiplier methods is ever called during reconstruction job, so Icache locality is good. See [sse2.c](https://github.com/zfsonlinux/zfs/blob/master/module/zfs/vdev_raidz_math_sse2.c#L256-L272) This would be only worth it if performance gain is significant enough (better than avx2 in last rec_ methods). However, there's no way of knowing how future hardware will handle shuffles...

**Comment by rdolbeau** [comment]
Rebased.

**Comment by ironMann** [comment]
@behlendorf @rdolbeau If it's not too much to ask, I would prefer to see this as two commits and PRs.
I can test only avx512f implementation, and I would rather leave 'bw' variant for when/if server CPUs with avx512bw are available. Their timeline keeps slipping...

**Comment by behlendorf** [comment]
That would be best, then we could at least test and merge the avx512f implementation.

**Comment by jxiong** [comment]
@behlendorf I'll try to find a node to test avx512bw.

**Comment by rdolbeau** [comment]
@ironMann avx512bw (& @behlendorf both) can be tested in userland with the Intel SDE. I grant you it doesn't validate the avx512bw code in-kernel (there could, theoretically, be a corner-case difference in the handling of registers :-/ ) Dunno how to split the commit in two w/o a lot of copy-paste. Would it be possible to simply commit a runtime-disabled avx512bw until it has been tested on the hardware? Whomever get the hardware first can then test it trivially.

@jxiong Testing on Skylake server (SKX) would be great, but double-check your NDA since I wouldn't be surprised if it prohibited you from communicating anything about SKX, including the fact you have access to SKX...

**Comment by behlendorf** [comment]
> commit a runtime-disabled avx512bw

I'd be OK with this if we can't get it validated first.  Although it may be easier and safer to disable it all at compile time.  We don't want people enabling this until it's tested on real hardware.  Anyone doing this testing can easily remove the compile time disable.

**Comment by rdolbeau** [comment]
@behlendorf Disabled AVX512BW.

**Comment by behlendorf** [comment]
@rdolbeau thanks for disabling avx512bw for the moment.  @ironMann @jxiong have either you gotten a chance to test the avx512f implementation?

**Comment by jxiong** [comment]
@behlendorf @rdolbeau I have got a skylake set up for the testing. However, when I was trying to figure out if it can support avx512bw. This is what I got from ```cpuid```:

```
   extended feature flags (7):
      FSGSBASE instructions                    = true
      IA32_TSC_ADJUST MSR supported            = true
      BMI instruction                          = true
      HLE hardware lock elision                = true
      AVX2: advanced vector extensions 2       = true
      SMEP supervisor mode exec protection     = true
      BMI2 instructions                        = true
      enhanced REP MOVSB/STOSB                 = true
      INVPCID instruction                      = true
      RTM: restricted transactional memory     = true
      QM: quality of service monitoring        = true
      deprecated FPU CS/DS                     = true
      intel memory protection extensions       = true
      AVX512F: AVX-512 foundation instructions = true
      RDSEED instruction                       = true
      ADX instructions                         = true
      SMAP: supervisor mode access prevention  = true
      Intel processor trace                    = true
      AVX512PF: prefetch instructions          = false
      AVX512ER: exponent & reciprocal instrs   = false
      AVX512CD: conflict detection instrs      = true
      SHA instructions                         = false
      PREFETCHWT1                              = false
```

Unfortunately I didn't find anything about 512bw, and this is the output from ```cpuinfo```:

```
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb pln pts dtherm tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx avx512f rdseed adx smap clflushopt avx512cd xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc
```
It looks like it doesn't support ```avx512bw``` or I have missed something here.

**Comment by rdolbeau** [comment]
@jxiong A Skylake with AVX512F should have AVX512BW as well, even though 'cpuid' doesn't yet support it. You can easily check in userland by recompiling SPL+ZFS (checking than you tools are recent enough to support AVX512*, look at the output of 'configure') and running "cmd/raidz_test/raidz_test -S -v". If AVX512BW is compiled in (disable commit c371307 !) , it should be 'supported' and shown as such in the output:
```
        Testing [avx512bw] implementation...[SUPPORTED]
                Testing method [gen_p] ...[PASS]
```

**Comment by jxiong** [comment]
@rdolbeau yes, it worked. This is what it printed when I ran the ```raidz_test```:

```
# ./cmd/raidz_test/raidz_test -S -v |grep avx512bw
	Testing [avx512bw] implementation...[SUPPORTED]
```

And the corresponding result from ```zfs_config.h```:
```
/* Define if host toolchain supports AVX512BW */
#define HAVE_AVX512BW 1
```
However, after compile and install zfs modules, I didn't find it from ```zfs_vdev_raidz_impl```:
```
# cat /sys/module/zfs/parameters/zfs_vdev_raidz_impl 
[fastest] original scalar sse2 ssse3 avx2 avx512f 
```

**Comment by rdolbeau** [comment]
@jxiong If it works in userland the hardware has the proper bits in the cpuid registers; however your kernel may not support all the features - ZFS needs X86_FEATURE_AVX512BW to be defined, which I believe was introduced in 4.6.

**Comment by jxiong** [comment]
@rdolbeau @behlendorf I compiled and installed a 4.9-rc3 kernel, and performed the following test:
1. make sure raidz impl is set to avx512bw;
2. create a raidz2 pool with 6 disks;
3. copy some big files into the pool;
4. offline two disks;
5. run ```zpool scrub```;
6. remount fs and compare files;

The test passed.

**Comment by rdolbeau** [comment]
@jxiong Excellent thank you :-)

**Comment by rlaager** [comment]
@jxiong If you still have everything setup, can you show your performance numbers: `cat /proc/spl/kstat/zfs/vdev_raidz_bench`

**Comment by jxiong** [comment]
@rlaager 
```
# cat /proc/spl/kstat/zfs/vdev_raidz_bench 
15 0 0x01 -1 0 1729626663499 47345243705423
implementation   gen_p           gen_pq          gen_pqr         rec_p           rec_q           rec_r           rec_pq          rec_pr          rec_qr          rec_pqr         
original         375807960       110323477       41848865        439405167       81989552        11252967        51296591        6963015         6884779         4793657         
scalar           258788159       151163676       66648866        302735160       73985482        55408697        60546216        47173766        27832270        22304402        
sse2             1149455456      517482566       254017817       1089754766      293236368       207339311       233734219       175748701       111245325       86096820        
ssse3            1149581672      615694539       261052300       1100237707      363203340       243171195       312296365       217029490       156311529       147407843       
avx2             1965007734      1180728235      501267621       1916004285      712349844       478576346       594288869       420565972       304037840       292498017       
avx512f          2711606855      1499955363      526771083       2782138649      773589448       521069416       579193250       376828360       260260742       233488124       
avx512bw         2724375124      1499781792      528646174       2802979712      1025546110      628584337       1014445829      584524392       388882295       367237664       
fastest          avx512bw        avx512f         avx512bw        avx512bw        avx512bw        avx512bw        avx512bw        avx512bw        avx512bw        avx512bw  
```

**Comment by rdolbeau** [comment]
@jxiong Thank you for the numbers. I assume frequency is quite low on this pre-release part, since the absolute numbers for up to AVX2 are about 1/3 of my 3.6 Ghz (Turbo) desktop Haswell.

But I'm quite happy with the improvements using AVX512BW over AVX512F or AVX2 for the rec_* results :-)

**Comment by behlendorf** [comment]
@jxiong thank you for making the time to build and test this on a real hardware!  Then this is ready to merge with both implementations enabled.  @rdolbeau can you remove commit c371307 from this PR so it can be merged.  @ironMann @jxiong can I add a reviewws-by tag for each of you when when merging this?

**Comment by rdolbeau** [comment]
@behlendorf Disabling commit removed

**Comment by ironMann** [comment]
@behlendorf  Everything seem to be working on my 512f machine. LGTM. You can go ahead and my reviewed-by.

@rdolbeau Once merged this will break ABD stack. What need's to be changed is final parametrization of methods, if I remember correctly. E.g. see [avx2.c](https://github.com/dpquigl/zfs/blob/3bbba33214d3eb2c2a61f68581dc390fed16d385/module/zfs/vdev_raidz_math_avx2.c#L338-L418) for #5135. The same goes for ARM versions...

**Comment by behlendorf** [comment]
@ironMann @rdolbeau after merging this PR getting a new one open to tackle the remaining work to get the methods parameterized would be very helpful.  If this could be done independently of the ABD work that would be ideal.  I'm sure @dpquigl would appreciate it, all the recent changes in this area have caused some complications.

**Pull request merged**